pytest_mh/utils/tc.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import annotations
+
+from .. import MultihostHost, MultihostUtility
+from ..ssh import SSHLog, SSHProcessError
+
+__all__ = ["LinuxTrafficControl"]
+
+
+class LinuxTrafficControl(MultihostUtility):
+    """
+    Perform traffic control operations on remote host.
+
+    All changes are automatically reverted when a test is finished.
+    """
+
+    def __init__(self, host: MultihostHost) -> None:
+        """
+        :param host: Remote host instance.
+        :type host: MultihostHost
+        """
+        super().__init__(host)
+        self.interface: str = ""
+        self.__restore: set[str] = set()
+        self.__host: MultihostHost = host
+
+    def setup_when_used(self):
+        """
+        Setup traffic control configuration.
+
+        :meta private:
+        """
+        super().setup_when_used()
+
+        # Let's find out the interface that connects the systems
+        result = self.host.ssh.run(f"dig +short {self.__host.hostname}", log_level=SSHLog.Error)
+        try:
+            result = self.host.ssh.run(f"ip -o address | grep {result.stdout}", log_level=SSHLog.Error)
+            self.interface = result.stdout.split(" ")[1]
+        except SSHProcessError:
```

Comment:
Is any error ok? Don't you want just check for the missing command?

You would be worth to log "ip command is missing, default interface is not set" kind of message.

dig may return multiple ip addresses, how do we handle that?

Reply:
> Is any error ok? Don't you want just check for the missing command?

Yes, how can I do that?

> dig may return multiple ip addresses, how do we handle that?

I'd default to the first one provided to dig, which is what I'm currently doing. If the test cases wants something else they can provide their selection in the interface when calling `add_delay()`.

Reply:
> > Is any error ok? Don't you want just check for the missing command?
> 
> Yes, how can I do that?

bash returns 127 and `bash: a: command not found`, we could check for this. But other shell may give other rc and the string may be translated. I think that we can really just omit catching the exception. If an error occurs, its reason will be fully visible in the logs.

> > dig may return multiple ip addresses, how do we handle that?
> 
> I'd default to the first one provided to dig, which is what I'm currently doing. If the test cases wants something else they can provide their selection in the interface when calling `add_delay()`.

I need to check how tc works exactly. Ideally, we don't want to slow down ssh connection. And the use case is to slow down connection to other host. Maybe the code in setup should be available as a method to return interface for specific hostname? And probably list of interfaces would be a better choice, because SSSD can pick any available IP address, not just the first one. So in case there are multiple IP addresses, SSSD would be slow only on one of them.

Reply:
> I need to check how tc works exactly. Ideally, we don't want to slow down ssh connection. And the use case is to slow down connection to other host. Maybe the code in setup should be available as a method to return interface for specific hostname? And probably list of interfaces would be a better choice, because SSSD can pick any available IP address, not just the first one. So in case there are multiple IP addresses, SSSD would be slow only on one of them.

IIUC, your proposal would be to create a method to return the available interfaces to a host, and let each test decide which one to choose. Right? That seems problematic in the sense that the person writing the test case may not know which interface to choose.

Reply:
something in the sense `delay(hostname, time)` instead of `delay(interface, time)` (or actually support both)

I need to get familiar with tc before making more suggestions. Quick search show that it should be possible [to delay only traffic to specific IP](https://serverfault.com/questions/389290/using-tc-to-delay-packets-to-only-a-single-ip-address). So `delay(hostname, time)` should work like:

```
devs = []
list_of_ips = dig hostname
for ip in list_of_ips:
     devs += (ip, dev) // find dev for ip
for ip, dev in devs
    tc ... delay communication to ip on dev by time
```

Reply:
It's better explained here https://dev.to/yugabyte/testing-an-application-with-a-network-delay-13eh, or at least I understood it better.

I'm working on a solution to this.

Reply:
Taking into account the use case I change the interface to `add_delay(hostname, time)`. The old interface doesn't make sense for that use case and it could be problematic. If at some point we need it we can add it.